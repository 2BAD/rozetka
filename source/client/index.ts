import got from 'got'
import { logRequest, logResponse } from './hooks/debug.js'
import { updateSearchParams } from './hooks/searchParams.js'

export const rozetka = got.extend({
  headers: {
    'user-agent': '2bad/rozetka'
  },
  hooks: {
    beforeRequest: [logRequest, updateSearchParams],
    afterResponse: [logResponse]
  },
  // only `lang` parameter is needed, the rest is just to simulate the frontend
  searchParams: new URLSearchParams([
    ['front-type', 'xl'],
    ['country', 'UA'],
    ['lang', 'ru']
  ])
})

export const catalog = rozetka.extend({
  prefixUrl: 'https://xl-catalog-api.rozetka.com.ua/',
  searchParams: new URLSearchParams([
    ['with_groups', '1'],
    ['with_docket', '1'],
    ['goods_group_href', '1']
  ])
})
