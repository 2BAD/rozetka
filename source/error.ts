import { z } from 'zod'

export const errorMap: z.ZodErrorMap = (error, ctx) => {
  if (error.code === z.ZodIssueCode.custom) {
    switch (error.params?.code) {
      case 'empty_response':
        return { message: `Bad response. The item you requested is not available.` }
    }
  }

  // fall back to default message
  return { message: ctx.defaultError }
}
