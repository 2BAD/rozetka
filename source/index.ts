export * as catalog from './catalog'
export * as product from './product'
