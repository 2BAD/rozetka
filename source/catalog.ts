import { catalog, common } from './client/index'
import { buildParams } from './client/utils/params'

export const categories = () => common.get('v2/fat-menu/full')

export const details = (categoryId: number) =>
  catalog.get('v4/categories/get', buildParams([['id', categoryId.toString()]]))

// @todo should paginate automatically
export const ids = (categoryId: number) => {
  return catalog.get(
    'v4/goods/get',
    buildParams([['category_id', categoryId.toString()]])
  )
}
