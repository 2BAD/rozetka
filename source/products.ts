import { catalog, common, product } from './client/index'
import { buildParams } from './client/utils/params'
import { Characteristics, Description } from './types/product'

/**
 * Create a common set of parameters for most of the api calls
 */
const genericParams = (id: number) => buildParams([['goodsId', id.toString()]])

export const details = (productId: number | number[]) =>
  catalog.get(
    'v4/goods/getDetails',
    buildParams([
      ['with_groups', '1'],
      ['with_docket', '1'],
      ['goods_group_href', '1'],
      ['product_ids', productId.toString()]
    ])
  )

export const description = (productId: number): Promise<Description> =>
  product
    .get('v4/goods/get-goods-description', genericParams(productId))
    .then((r) => Description.parse(r))

//@todo check if there can by more than one element in response array
export const characteristics = (productId: number): Promise<Characteristics> =>
  product
    .get('v4/goods/get-characteristic', genericParams(productId))
    .then((r) => Characteristics.parse(r))

export const tags = (productId: number) =>
  product.get('v4/marketing/get-tags', genericParams(productId))

export const promotions = (productId: number) =>
  product.get('v4/marketing/get-promotions', genericParams(productId))

export const offers = (productId: number) =>
  product.get('v4/marketing/get-super-offer', genericParams(productId))

export const relatedProducts = (productId: number) =>
  product.get('v4/goods/get-related', genericParams(productId))

export const price = (productId: number | number[]) => {
  const paramName = Array.isArray(productId) ? 'ids' : 'id'
  // trailing slash is necessary here
  return common.get(
    'v2/goods/get-price/',
    buildParams([[paramName, productId.toString()]])
  )
}

try {
  let c = await characteristics(6802844)

  console.log(c)
} catch (error) {
  console.log(error)
}
