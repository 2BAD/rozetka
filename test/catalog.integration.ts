import { catalog } from '../source/index'

jest.setTimeout(20000)

describe('catalog', () => {
  const CATEGORY_ID = 4649166

  it('categories should match snapshot', async () => {
    expect.assertions(1)
    const categories = await catalog.categories()
    expect(Object.keys(categories)).toMatchSnapshot()
  })

  it('children should match snapshot', async () => {
    expect.assertions(1)
    const children = await catalog.children(CATEGORY_ID)
    expect(children).toMatchSnapshot()
  })

  it('details should match snapshot', async () => {
    expect.assertions(1)
    const details = await catalog.details(CATEGORY_ID)
    expect(details).toMatchSnapshot()
  })

  it('ids should match snapshot', async () => {
    expect.assertions(1)
    const ids = await catalog.ids(CATEGORY_ID)
    expect(ids.sort((a, b) => a - b)).toMatchSnapshot()
  })
})
