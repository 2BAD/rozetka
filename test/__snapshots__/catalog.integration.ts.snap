// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`catalog categories should match snapshot 1`] = `
Array [
  "2416",
  "3361",
  "4306",
  "5300",
  "6700",
  "7806",
  "8261",
  "9017",
  "10515",
  "12258",
  "13224",
  "14127",
  "14939",
  "15422",
  "15954",
  "17191",
  "18421",
  "18931",
  "18934",
]
`;

exports[`catalog children should match snapshot 1`] = `
Object {
  "children": Array [],
  "current": Object {
    "allow_index_three_parameters": false,
    "count": "202",
    "href": "https://rozetka.com.ua/dgin/c4649166/",
    "id": 4649166,
    "is_rozetka_top": true,
    "left_key": 8037,
    "level": 4,
    "mpath": ".4626923.4594201.4594292.",
    "name": "dgin",
    "parent_id": 4594292,
    "right_key": 8038,
    "root_id": 4626923,
    "rz_mpath": ".4626923.4594201.4594292.4649166.",
    "status": "active",
    "subdomain": null,
    "title": "Джин",
    "titles_mode": "inherited",
  },
  "hasFilter": false,
  "is_need_street_id": false,
  "tile_config": Object {
    "possible_templates": Array [
      "default",
      "wide",
    ],
    "template": "default",
    "template_view": "default",
  },
}
`;

exports[`catalog details should match snapshot 1`] = `
Object {
  "category": Object {
    "allow_index_three_parameters": false,
    "href": "https://rozetka.com.ua/dgin/c4649166/",
    "id": 4649166,
    "is_rozetka_top": true,
    "left_key": 8037,
    "level": 4,
    "mpath": ".4626923.4594201.4594292.",
    "name": "dgin",
    "parent_id": 4594292,
    "right_key": 8038,
    "root_id": 4626923,
    "rz_mpath": ".4626923.4594201.4594292.4649166.",
    "status": "active",
    "title": "Джин",
    "titles_mode": "inherited",
  },
  "is_need_street_id": false,
  "is_shop_category": false,
  "parents": Array [
    Object {
      "allow_index_three_parameters": false,
      "href": "https://rozetka.com.ua/alkoholnie-napitki-i-produkty/c4626923/",
      "id": 4626923,
      "is_rozetka_top": false,
      "left_key": 8004,
      "level": 1,
      "mpath": ".",
      "name": "alkoholnie-napitki-i-produkty",
      "parent_id": 0,
      "right_key": 8501,
      "rz_mpath": ".4626923.",
      "status": "active",
      "title": "Алкогольные напитки и продукты питания",
      "titles_mode": "inherited",
    },
    Object {
      "allow_index_three_parameters": false,
      "href": "https://rozetka.com.ua/alkogolnie-napitki-konfety/c4594201/",
      "id": 4594201,
      "is_rozetka_top": false,
      "left_key": 8013,
      "level": 2,
      "mpath": ".4626923.",
      "name": "alkogolnie-napitki-konfety",
      "parent_id": 4626923,
      "right_key": 8056,
      "rz_mpath": ".4626923.4594201.",
      "status": "active",
      "title": "Алкогольные напитки",
      "titles_mode": "inherited",
    },
    Object {
      "allow_index_three_parameters": false,
      "href": "https://rozetka.com.ua/krepkie-napitki/c4594292/",
      "id": 4594292,
      "is_rozetka_top": false,
      "left_key": 8022,
      "level": 3,
      "mpath": ".4626923.4594201.",
      "name": "krepkie-napitki",
      "parent_id": 4594201,
      "right_key": 8045,
      "rz_mpath": ".4626923.4594201.4594292.",
      "status": "active",
      "title": "Крепкие напитки",
      "titles_mode": "inherited",
    },
  ],
  "redirect_to": null,
  "show_brands_slider": false,
  "subdomain": null,
  "tile_config": Object {
    "possible_templates": Array [
      "default",
      "wide",
    ],
    "template": "default",
    "template_view": "default",
  },
}
`;

exports[`catalog ids should match snapshot 1`] = `
Array [
  4727702,
  4727828,
  4727870,
  4727891,
  4729473,
  4729655,
  4729753,
  4977706,
  4977811,
  4977867,
  5704761,
  5705499,
  5719503,
  5720079,
  5720193,
  5720271,
  5824281,
  5887449,
  5888247,
  5908917,
  5909175,
  5909253,
  5909349,
  7117144,
  7117999,
  7118174,
  7118269,
  7118409,
  7770920,
  7772649,
  9499389,
  9524799,
  11111391,
  12156369,
  12318350,
  12827600,
  13816973,
  13817022,
  13839373,
  13839422,
  16998038,
  17170597,
  18249896,
  19682782,
  20302898,
  20306370,
  24672711,
  25108513,
  25446329,
  33211767,
  33211895,
  36004944,
  36010608,
  36019656,
  39077576,
  40446296,
  44856936,
  46244464,
  46270888,
  49257118,
  50015654,
  50015710,
  54697632,
  54698376,
  54698904,
  54699270,
  54701856,
  56284026,
  56294118,
  58608940,
  58609324,
  61826651,
  61846216,
  61855064,
  65418899,
  71617564,
  71617594,
  71617612,
  71618830,
  73053561,
  73789407,
  74007651,
  74012589,
  74825886,
  74869128,
  80784117,
  85351344,
  88622019,
  95267176,
  96311272,
  107396442,
  108644742,
  117263239,
  118810753,
  129736913,
  129736927,
  129736934,
  129736941,
  129737011,
  129737137,
  135538682,
  135814769,
  136543175,
  144120608,
  144152276,
  144153242,
  154188021,
  154189547,
  154386793,
  156748999,
  159467351,
  159509365,
  159600148,
  173040806,
  173056696,
  178978948,
  183916958,
  189605389,
  192705297,
  204746383,
  204749827,
  205979749,
  207321475,
  214807315,
  222200107,
  223083631,
  230005399,
  231139645,
  232680421,
  235887955,
  237907963,
  248933359,
  248957431,
  248966737,
  252843286,
  252851256,
  256331251,
  258080351,
  263382626,
  263946666,
  264193491,
  264681641,
  276349483,
  276350883,
  277908968,
  278352688,
  283887778,
  283887783,
  283887808,
  283887833,
  283887863,
  283887868,
  283887873,
  283887878,
  283887883,
  283887908,
  283887963,
  283887983,
  295915648,
  297468133,
  299127773,
  299140418,
  303656623,
  304006233,
  304006238,
  304006243,
  304006248,
  304006253,
  304006258,
  304307178,
  305366608,
  309550758,
  309550788,
  309550803,
  309550833,
  309550863,
  311954828,
  313415164,
  313425463,
  317923027,
  317947303,
  317947471,
  319051090,
  320083417,
  320083420,
  320277457,
  323221765,
  323477965,
  326562571,
  327346633,
  327928651,
  328422934,
  328422937,
  328422985,
  328423003,
  328645642,
  329253343,
  329253358,
  329253553,
  329253829,
  329253973,
  330685423,
]
`;
